File: src/main/bridge/EditorPayloadCodec.ts
import type {
  EditorCommand,
  EditorEntity,
  EditorEvent,
  EditorLevelSettings,
  EditorSnapshot,
  ProjectTransform,
  ProjectVector3,
} from '../../types/EditorRuntime.js';
import { PayloadReader, PayloadWriter, malformed } from './PayloadIO.js';

const MAX_ENTITIES = 100_000;
const MAX_EVENTS = 1_024;
const commandKinds = {
  setSelection: 1,
  renameProject: 2,
  updateTransform: 3,
  renameEntity: 4,
  setEntityLayer: 5,
  setEntityState: 6,
  updateTransforms: 7,
  undo: 8,
  redo: 9,
  deleteEntities: 10,
  duplicateEntities: 11,
  copyEntities: 12,
  pasteEntities: 13,
  updateLevelSettings: 14,
  updateSplinePoints: 15,
} as const;
const eventKinds: Record<number, EditorEvent['kind']> = {
  1: 'projectOpened', 2: 'projectChanged', 3: 'selectionChanged', 4: 'projectSaved',
  5: 'recoveryWritten', 6: 'diagnosticRaised', 7: 'projectClosed',
};
const severities: Record<number, EditorSnapshot['diagnostics'][number]['severity']> = {
  1: 'info', 2: 'warning', 3: 'error',
};
const geometryKinds: Record<number, NonNullable<EditorEntity['geometry']>['kind']> = {
  1: 'cuboid', 2: 'spline', 3: 'area', 4: 'sphere', 5: 'cylinder', 6: 'pill', 7: 'grindPath',
  8: 'directionalLight', 9: 'pointLight', 10: 'environmentSample', 11: 'environmentTransition',
  12: 'camera', 13: 'ambientSound',
};

export function encodeEditorOpenRequest(projectPath: string, catalogRootPath: string, autosaveSeconds: number): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(projectPath);
  writer.writeString(catalogRootPath);
  writer.writeUInt32(autosaveSeconds);
  return writer.toBuffer();
}

export function encodeEditorCommand(value: EditorCommand): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.id);
  writer.writeUInt32(commandKinds[value.kind]);
  writeStrings(writer, value.entityIds);
  writer.writeBoolean(value.kind === 'updateTransform');
  if (value.kind === 'updateTransform') writeTransform(writer, value.transform);
  const transforms = value.kind === 'updateTransforms' ? value.transforms : [];
  writer.writeUInt32(transforms.length);
  transforms.forEach((update) => {
    writer.writeString(update.entityId);
    writeTransform(writer, update.transform);
  });
  const hasText = value.kind === 'renameProject' || value.kind === 'renameEntity' || value.kind === 'setEntityLayer';
  writer.writeBoolean(hasText);
  if (hasText) writer.writeString(value.text);
  writer.writeBoolean(value.kind === 'setEntityState');
  if (value.kind === 'setEntityState') {
    writeOptionalBoolean(writer, value.state.hidden);
    writeOptionalBoolean(writer, value.state.disabled);
    writeOptionalBoolean(writer, value.state.locked);
  }
  writer.writeBoolean(value.kind === 'updateLevelSettings');
  if (value.kind === 'updateLevelSettings') writeLevelSettings(writer, value.levelSettings);
  writer.writeBoolean(value.kind === 'updateSplinePoints');
  if (value.kind === 'updateSplinePoints') {
    writer.writeUInt32(value.points.length);
    value.points.forEach((point) => {
      writer.writeFloat32(point.x);
      writer.writeFloat32(point.y);
      writer.writeFloat32(point.z);
      writer.writeFloat32(point.w);
    });
  }
  return writer.toBuffer();
}

export function encodeEditorEventRequest(afterSequence: number, limit: number): Buffer {
  const writer = new PayloadWriter();
  writer.writeUInt64(afterSequence);
  writer.writeUInt32(limit);
  return writer.toBuffer();
}

export function decodeEditorSnapshot(payload: Uint8Array): EditorSnapshot {
  const reader = new PayloadReader(payload);
  const projectPath = reader.readString();
  const projectId = reader.readString();
  const projectName = reader.readString();
  const target = {
    game: reader.readString(), region: reader.readString(),
    revision: reader.readString(), bakeProfile: reader.readString(),
  };
  const baseLevel = {
    game: reader.readString(), region: reader.readString(), revision: reader.readString(),
    level: reader.readUInt32(), sourceFingerprint: reader.readString(), missingAssetCount: reader.readUInt32(),
  };
  const levelSettings = reader.readBoolean() ? readLevelSettings(reader) : undefined;
  const entities = readList(reader, MAX_ENTITIES, () => readEntity(reader));
  const selection = readStrings(reader, MAX_ENTITIES);
  const isDirty = reader.readBoolean();
  const migrationPending = reader.readBoolean();
  const canUndo = reader.readBoolean();
  const canRedo = reader.readBoolean();
  const canPaste = reader.readBoolean();
  const lastEventSequence = reader.readUInt64();
  const capabilities = reader.readStrings();
  const tools = readList(reader, 64, () => ({
    id: reader.readString(), label: reader.readString(), capability: reader.readString(),
  }));
  const diagnostics = readList(reader, 100, () => ({
    code: reader.readString(), severity: enumValue(severities, reader.readUInt32(), 'diagnostic severity'),
    message: reader.readString(),
  }));
  reader.complete();
  return {
    projectPath, projectId, projectName, target, baseLevel, levelSettings, entities, selection, isDirty,
    migrationPending, canUndo, canRedo, canPaste, lastEventSequence, capabilities, tools, diagnostics,
  };
}

export function decodeEditorEvents(payload: Uint8Array): EditorEvent[] {
  const reader = new PayloadReader(payload);
  const values = readList(reader, MAX_EVENTS, () => ({
    sequence: reader.readUInt64(),
    createdUnixMilliseconds: reader.readUInt64(),
    kind: enumValue(eventKinds, reader.readUInt32(), 'event kind'),
    commandId: reader.readBoolean() ? reader.readString() : undefined,
    entityIds: readStrings(reader, MAX_ENTITIES),
    message: reader.readBoolean() ? reader.readString() : undefined,
  }));
  reader.complete();
  return values;
}

function readEntity(reader: PayloadReader): EditorEntity {
  const value: EditorEntity = {
    id: reader.readString(), name: reader.readString(), layer: reader.readString(), transform: readTransform(reader),
    state: {
      dirty: false, hidden: false, disabled: false, locked: false, readOnly: false,
      invalid: false, missingAsset: false,
    },
  };
  if (reader.readBoolean()) value.asset = { id: reader.readString(), kind: reader.readString() };
  if (reader.readBoolean()) value.provenance = {
    game: reader.readString(), level: reader.readUInt32(), section: reader.readString(), sourceIndex: reader.readUInt32(),
  };
  if (reader.readBoolean()) value.sourceClassId = reader.readUInt32();
  if (reader.readBoolean()) value.geometry = {
    kind: enumValue(geometryKinds, reader.readUInt32(), 'geometry kind'),
    points: readList(reader, MAX_ENTITIES, () => ({
      x: reader.readFloat32(), y: reader.readFloat32(), z: reader.readFloat32(), w: reader.readFloat32(),
    })),
  };
  value.state = {
    dirty: reader.readBoolean(),
    hidden: reader.readBoolean(),
    disabled: reader.readBoolean(),
    locked: reader.readBoolean(),
    readOnly: reader.readBoolean(),
    invalid: reader.readBoolean(),
    missingAsset: reader.readBoolean(),
  };
  return value;
}

function writeOptionalBoolean(writer: PayloadWriter, value: boolean | undefined): void {
  writer.writeBoolean(value !== undefined);
  if (value !== undefined) writer.writeBoolean(value);
}

function writeLevelSettings(writer: PayloadWriter, value: EditorLevelSettings): void {
  value.backgroundColor.forEach((channel) => writer.writeUInt32(channel));
  value.fogColor.forEach((channel) => writer.writeUInt32(channel));
  writer.writeFloat32(value.fogNearDistance);
  writer.writeFloat32(value.fogFarDistance);
  writer.writeFloat32(value.fogNearIntensity);
  writer.writeFloat32(value.fogFarIntensity);
}

function readLevelSettings(reader: PayloadReader): EditorLevelSettings {
  return {
    backgroundColor: [reader.readUInt32(), reader.readUInt32(), reader.readUInt32()],
    fogColor: [reader.readUInt32(), reader.readUInt32(), reader.readUInt32()],
    fogNearDistance: reader.readFloat32(),
    fogFarDistance: reader.readFloat32(),
    fogNearIntensity: reader.readFloat32(),
    fogFarIntensity: reader.readFloat32(),
  };
}

function writeTransform(writer: PayloadWriter, value: ProjectTransform): void {
  writeVector(writer, value.position);
  writer.writeFloat32(value.rotation.x);
  writer.writeFloat32(value.rotation.y);
  writer.writeFloat32(value.rotation.z);
  writer.writeFloat32(value.rotation.w);
  writeVector(writer, value.scale);
}

function readTransform(reader: PayloadReader): ProjectTransform {
  return {
    position: readVector(reader),
    rotation: { ...readVector(reader), w: reader.readFloat32() },
    scale: readVector(reader),
  };
}

function writeVector(writer: PayloadWriter, value: ProjectVector3): void {
  writer.writeFloat32(value.x);
  writer.writeFloat32(value.y);
  writer.writeFloat32(value.z);
}

function readVector(reader: PayloadReader): ProjectVector3 {
  return { x: reader.readFloat32(), y: reader.readFloat32(), z: reader.readFloat32() };
}

function writeStrings(writer: PayloadWriter, values: string[]): void {
  if (values.length > MAX_ENTITIES) malformed('Entity ID list exceeds item limit');
  writer.writeUInt32(values.length);
  values.forEach((value) => writer.writeString(value));
}

function readStrings(reader: PayloadReader, maximum: number): string[] {
  return readList(reader, maximum, () => reader.readString());
}

function readList<T>(reader: PayloadReader, maximum: number, read: () => T): T[] {
  const count = reader.readUInt32();
  if (count > maximum) malformed('List exceeds item limit');
  return Array.from({ length: count }, read);
}

function enumValue<T extends string>(values: Record<number, T>, index: number, name: string): T {
  const value = values[index];
  if (!value) malformed(`Unknown editor ${name}`);
  return value;
}

File: src/Forge.Host/Bridge/EditorPayloadCodec.cs
using Forge.Host.Domain;

namespace Forge.Host.Bridge;

internal readonly record struct EditorOpenRequest(string ProjectPath, string CatalogRootPath, uint AutosaveSeconds);
internal readonly record struct EditorEventRequest(ulong AfterSequence, uint Limit);

internal static class EditorPayloadCodec
{
    private const uint MaxEntities = 100_000;
    private const uint MaxEvents = 1_024;

    public static EditorOpenRequest DecodeOpenRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new EditorOpenRequest(reader.ReadString(), reader.ReadString(), reader.ReadUInt32());
        reader.Complete();
        return value;
    }

    public static EditorEventRequest DecodeEventRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new EditorEventRequest(reader.ReadUInt64(), reader.ReadUInt32());
        reader.Complete();
        if (value.AfterSequence > long.MaxValue) PayloadFormat.Malformed("Invalid editor event sequence");
        if (value.Limit is 0 or > MaxEvents) PayloadFormat.Malformed("Invalid editor event limit");
        return value;
    }

    public static EditorCommand DecodeCommand(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var id = reader.ReadString();
        var kind = (EditorCommandKind)reader.ReadUInt32();
        var entities = ReadEntityIds(ref reader);
        var transform = reader.ReadBoolean() ? ReadTransform(ref reader) : null;
        var transformCount = reader.ReadUInt32();
        if (transformCount > MaxEntities) PayloadFormat.Malformed("Transform list exceeds item limit");
        var transforms = new EditorTransformUpdate[transformCount];
        for (var index = 0; index < transforms.Length; index++)
            transforms[index] = new(EntityId.Parse(reader.ReadString()), ReadTransform(ref reader));
        var text = reader.ReadBoolean() ? reader.ReadString() : null;
        var state = reader.ReadBoolean() ? ReadStateChange(ref reader) : null;
        reader.Complete();
        return new(id, kind, entities, transform, text, state, transforms);
    }

    public static byte[] EncodeSnapshot(EditorSnapshot value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.ProjectPath);
        writer.WriteString(value.ProjectId.ToString());
        writer.WriteString(value.ProjectName);
        WriteTarget(writer, value.Target);
        WriteBaseLevel(writer, value.BaseLevel);
        if (value.Entities.Count > MaxEntities) PayloadFormat.Malformed("Entity list exceeds item limit");
        writer.WriteUInt32((uint)value.Entities.Count);
        foreach (var entity in value.Entities) WriteEntity(writer, entity);
        WriteEntityIds(writer, value.Selection);
        writer.WriteBoolean(value.IsDirty);
        writer.WriteBoolean(value.MigrationPending);
        writer.WriteBoolean(value.CanUndo);
        writer.WriteBoolean(value.CanRedo);
        writer.WriteBoolean(value.CanPaste);
        writer.WriteUInt64(checked((ulong)value.LastEventSequence));
        writer.WriteStrings(value.Capabilities.ToArray());
        writer.WriteUInt32((uint)value.Tools.Count);
        foreach (var tool in value.Tools)
        {
            writer.WriteString(tool.Id);
            writer.WriteString(tool.Label);
            writer.WriteString(tool.Capability);
        }
        writer.WriteUInt32((uint)value.Diagnostics.Count);
        foreach (var diagnostic in value.Diagnostics)
        {
            writer.WriteString(diagnostic.Code);
            writer.WriteUInt32((uint)diagnostic.Severity);
            writer.WriteString(diagnostic.Message);
        }
        return writer.ToArray();
    }

    public static byte[] EncodeEvents(IReadOnlyList<EditorEvent> values)
    {
        if (values.Count > MaxEvents) PayloadFormat.Malformed("Event list exceeds item limit");
        var writer = new PayloadWriter();
        writer.WriteUInt32((uint)values.Count);
        foreach (var value in values)
        {
            writer.WriteUInt64(checked((ulong)value.Sequence));
            writer.WriteUInt64(checked((ulong)value.CreatedUnixMilliseconds));
            writer.WriteUInt32((uint)value.Kind);
            writer.WriteBoolean(value.CommandId is not null);
            if (value.CommandId is not null) writer.WriteString(value.CommandId);
            WriteEntityIds(writer, value.EntityIds);
            writer.WriteBoolean(value.Message is not null);
            if (value.Message is not null) writer.WriteString(value.Message);
        }
        return writer.ToArray();
    }

    private static void WriteTarget(PayloadWriter writer, ProjectTargetProfile value)
    {
        writer.WriteString(value.Game);
        writer.WriteString(value.Region);
        writer.WriteString(value.Revision);
        writer.WriteString(value.BakeProfile);
    }

    private static void WriteBaseLevel(PayloadWriter writer, ProjectBaseLevel value)
    {
        writer.WriteString(value.Game);
        writer.WriteString(value.Region);
        writer.WriteString(value.Revision);
        writer.WriteUInt32(checked((uint)value.Level));
        writer.WriteString(value.SourceFingerprint);
        writer.WriteUInt32(checked((uint)value.MissingAssetCount));
    }

    private static void WriteEntity(PayloadWriter writer, EditorEntitySnapshot value)
    {
        writer.WriteString(value.EntityId.ToString());
        writer.WriteString(value.Name);
        writer.WriteString(value.Layer);
        WriteTransform(writer, value.Transform);
        writer.WriteBoolean(value.Asset is not null);
        if (value.Asset is not null)
        {
            writer.WriteString(value.Asset.Id.ToString());
            writer.WriteString(value.Asset.Kind.ToString());
        }
        writer.WriteBoolean(value.Provenance is not null);
        if (value.Provenance is not null)
        {
            writer.WriteString(value.Provenance.Game);
            writer.WriteUInt32(checked((uint)value.Provenance.Level));
            writer.WriteString(value.Provenance.Section);
            writer.WriteUInt32(checked((uint)value.Provenance.SourceIndex));
        }
        writer.WriteBoolean(value.SourceClassId is not null);
        if (value.SourceClassId is not null) writer.WriteUInt32(checked((uint)value.SourceClassId));
        writer.WriteBoolean(value.Geometry is not null);
        if (value.Geometry is not null)
        {
            writer.WriteUInt32((uint)value.Geometry.Kind);
            if (value.Geometry.Points.Count > MaxEntities) PayloadFormat.Malformed("Geometry point list exceeds item limit");
            writer.WriteUInt32((uint)value.Geometry.Points.Count);
            foreach (var point in value.Geometry.Points)
            {
                writer.WriteSingle(point.X);
                writer.WriteSingle(point.Y);
                writer.WriteSingle(point.Z);
                writer.WriteSingle(point.W);
            }
        }
        writer.WriteBoolean(value.State.Dirty);
        writer.WriteBoolean(value.State.Hidden);
        writer.WriteBoolean(value.State.Disabled);
        writer.WriteBoolean(value.State.Locked);
        writer.WriteBoolean(value.State.ReadOnly);
        writer.WriteBoolean(value.State.Invalid);
        writer.WriteBoolean(value.State.MissingAsset);
    }

    private static EditorEntityStateChange ReadStateChange(ref PayloadReader reader) => new(
        reader.ReadBoolean() ? reader.ReadBoolean() : null,
        reader.ReadBoolean() ? reader.ReadBoolean() : null,
        reader.ReadBoolean() ? reader.ReadBoolean() : null);

    private static void WriteTransform(PayloadWriter writer, ProjectTransform value)
    {
        WriteVector(writer, value.Position);
        writer.WriteSingle(value.Rotation.X);
        writer.WriteSingle(value.Rotation.Y);
        writer.WriteSingle(value.Rotation.Z);
        writer.WriteSingle(value.Rotation.W);
        WriteVector(writer, value.Scale);
    }

    private static ProjectTransform ReadTransform(ref PayloadReader reader) => new(
        ReadVector(ref reader),
        new(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle()),
        ReadVector(ref reader));

    private static void WriteVector(PayloadWriter writer, ProjectVector3 value)
    {
        writer.WriteSingle(value.X);
        writer.WriteSingle(value.Y);
        writer.WriteSingle(value.Z);
    }

    private static ProjectVector3 ReadVector(ref PayloadReader reader) =>
        new(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());

    private static void WriteEntityIds(PayloadWriter writer, IReadOnlyList<EntityId> values)
    {
        if (values.Count > MaxEntities) PayloadFormat.Malformed("Entity ID list exceeds item limit");
        writer.WriteUInt32((uint)values.Count);
        foreach (var value in values) writer.WriteString(value.ToString());
    }

    private static EntityId[] ReadEntityIds(ref PayloadReader reader)
    {
        var count = reader.ReadUInt32();
        if (count > MaxEntities) PayloadFormat.Malformed("Entity ID list exceeds item limit");
        var values = new EntityId[count];
        for (var index = 0; index < values.Length; index++) values[index] = EntityId.Parse(reader.ReadString());
        return values;
    }
}
